lib/facter/tpm/util.rb

```diff
@@ -0,0 +1,208 @@
+        require 'pty'
+
+        out = ''
+
+        begin
+          if owner_pass == 'well-known'
+            Puppet.debug('running tpm_getpubek using well-known option')
+            out = Facter::Core::Execution.execute('tpm_getpubek -z', :timeout => 15)
+          else
+            Puppet.debug('running tpm_getpubek')
+            ary = []
+            PTY.spawn(cmd) do |r,w,pid|
+              w.sync = true
+              begin
+                r.expect( /owner password/i, 15 ) { |s| w.puts owner_pass }
+                r.each { |line| ary << line }
+              rescue Errno::EIO
+                # just until the end of the IO stream
+              end
+              Process.wait(pid)
+            end
+            out = ary.join("\n")
+          end
+        rescue Facter::Core::Execution::ExecutionFailure
+          Puppet.debug('tpm_getpubek timed out!')
+          out = nil
+        end
+
+        out
+      end
+
+      # Get the pubek from tpm_getpubek when the TPM isn't owned
+      # @return [String] the output of the command, or nil if it times out
+      #
+      def get_pubek_unowned
+        begin
+          Puppet.debug('running tpm_getpubek')
+          status = Facter::Core::Execution.execute('tpm_getpubek', :timeout => 15)
+        rescue Facter::Core::Execution::ExecutionFailure
+          Puppet.debug('tpm_getpubek timed out!')
+          status = 'error: tpm_getpubek timed out'
+        end
+
+        status
+      end
+
+      # Get the output of tpm_version
+      # @return [String] the output of the command, or nil if it times out
+      #
+      def tpm_version
+        begin
+          Puppet.debug('running tpm_version')
+          status = Facter::Core::Execution.execute('tpm_version', :timeout => 15)
+        rescue Facter::Core::Execution::ExecutionFailure
+          Puppet.debug('tpm_version timed out!')
+          status = nil
+        end
+
+        status
+      end
+
+      # @return [Hash] the yaml output from `tpm_version`
+      #
+      # The output of tpm_version just so happens to be valid yaml, and
+      # it is parsed as such.
+      #
+      # Example output:
+      #  ```
+      #    TPM 1.2 Version Info:
+      #    Chip Version:        1.2.18.60
+      #    Spec Level:          2
+      #    Errata Revision:     3
+      #    TPM Vendor ID:       IBM
+      #    TPM Version:         01010000
+      #    Manufacturer Info:   49424d00
+      #  ```
+      #
+      def version
+        require 'yaml'
+
+        output = Hash.new
+        cmd_out = tpm_version
+
+        if cmd_out == "" or cmd_out.nil?
+          output['_status'] = 'Trousers is not running'
+          return output
+        else
+          version = YAML.load(cmd_out)
+
+          # Format keys in a way that Facter will like
+          begin
+            output = Hash[version.flatten[1].map{ |k,v| [k.downcase.gsub(/ /, '_'), v] }]
+          rescue NoMethodError
+            version.delete_if { |k,y| k =~ /Version Info/ }
```

Comment:
should be |k,v| ?

Reply:
What you call the variables is entirely up to you in Ruby